### PR TITLE
[exporter/elasticsearch] Flesh out encoding log records with `mapping.mode: ecs`

### PR DESCRIPTION
**Pull request by ycombinator**
**Description:** <Describe what has changed.>

This PR is a follow up to https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/31553.  In that PR, the foundations were laid for encoding log records with ECS fields when the user specifies `mapping.mode: ecs` in their exporter configuration.

This PR adds ECS conversions for several additional fields.

I expect there will be more followup PR(s) to add more conversions, especially ones that might not be straightforward.

**Link to tracking issue:** #29742 

**Testing:** Added unit tests

**Documentation:** The `mapping.mode: ecs` configuration setting is already [documented](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/exporter/elasticsearchexporter#configuration-options).

### CODE REVIEW

File: exporter/elasticsearchexporter/model.go
Comment (axw) [reply exporter/elasticsearchexporter/model.go:187]: Internally `pcommon.Map` is an array of key-value pairs (https://github.com/open-telemetry/opentelemetry-collector/blob/f0473ca0a55293b92374cc77e0275026537342da/pdata/pcommon/map.go#L26), so even if `conversionMap` is smaller we'd likely end up iterating `attrs` multiple times that way.

File: exporter/elasticsearchexporter/model.go
Comment (ycombinator) [reply exporter/elasticsearchexporter/model.go]: ++ done in 3ea4512c3e3451d73708fbed7d0ff2576cf3b304.

File: exporter/elasticsearchexporter/model.go
Comment (ycombinator) [reply exporter/elasticsearchexporter/model.go:192]: I did add an early return for the special case where `conversionMap` was empty: 47905d4e98ddccb243dcc3eadd2ac8ed1dc6e097.

File: exporter/elasticsearchexporter/model.go
Comment (axw) [inline exporter/elasticsearchexporter/model.go:95]: Should we be vetting the values here? ECS expects `os.platform` to have one of a restricted set of values -- https://www.elastic.co/guide/en/ecs/current/ecs-os.html#field-os-type

For iOS and Android, I think we should set `os.platform` from `os.name` ("Android" or "iOS") rather than `os.type` ("linux" or "darwin")

See https://github.com/elastic/apm-data/blob/0e072ceaadae2ef8f61d75c3a27c47cc5ce81257/input/otlp/metadata.go#L320-L338

**Review by axw — COMMENTED** [review]
Mostly LGTM, just a few comments so far.

File: exporter/elasticsearchexporter/model.go
Comment (felixbarny) [inline exporter/elasticsearchexporter/model.go:84]: I think what's missing is the translation of `k8s.*` SemConv resource attributes. See also https://github.com/elastic/apm-data/blob/0e072ceaadae2ef8f61d75c3a27c47cc5ce81257/input/otlp/metadata.go#L145-L165

File: exporter/elasticsearchexporter/model.go
Comment (carsonip) [reply exporter/elasticsearchexporter/model.go:47]: I imagine the closest one to event.received is observedTimestamp: https://opentelemetry.io/docs/specs/otel/logs/data-model/#field-observedtimestamp

I expect the receiver to set observedTimestamp, and the exporter to use it.

Comment (lahsivjar) [reply exporter/elasticsearchexporter/model.go:47]: > Perhaps we should omit the explicit mapping of event.received, and leave it as a dynamically mapped attribute? i.e. leave it to the receiver or some upstream client to set the event.received attribute.

+1 on this. The observed timestamp should be a fallback for the `@timestamp` field if the record doesn't have the timestamp (the PR already does that in `encodeLogTimestampECSMode` method).

File: exporter/elasticsearchexporter/model.go
Comment (ycombinator) [reply exporter/elasticsearchexporter/model.go:84]: Yep, I didn't include them as they're not really part of ECS.  But as per the off-PR discussion with @axw and you, I will include them now for practical reasons.

File: exporter/elasticsearchexporter/model.go
Comment (felixbarny) [reply exporter/elasticsearchexporter/model.go:47]: There's a bit of a difference when this logic gets executed. In APM Server, this is executed when receiving OTLP events. However, when doing it in the ES exporter it's set when exporting/sending events to ES. The difference is that there may be a delay in between receiving an exporting the events, especially if the events sit in a queue for a while. So to do it in a similar way as APM Server, we'd need to set `event.received` in the OTLP receiver. But that's not in the scope of improving the ES exporter.

Comment (ycombinator) [reply exporter/elasticsearchexporter/model.go:47]: Thanks for the clarification, @felixbarny.  Makes sense why the same logic does not apply in both scenarios.  Removed `event.received` setting code in d9e4f69a6e.

File: exporter/elasticsearchexporter/model.go
Comment (ycombinator) [reply exporter/elasticsearchexporter/model.go:84]: Added in 468043f049.

File: exporter/elasticsearchexporter/model.go
Comment (ycombinator) [reply exporter/elasticsearchexporter/model.go]: I moved this mapping out of this function and into package scope in bd90fbf4076586501b57b7799e6b12797cd30828 so, in the future, conversion code for other signals can use it too.

File: exporter/elasticsearchexporter/model.go
Comment (ycombinator) [reply exporter/elasticsearchexporter/model.go]: Done in 468043f0494859176c0fd89d3c1d6c62466068d2.

File: exporter/elasticsearchexporter/model.go
Comment (ycombinator) [reply exporter/elasticsearchexporter/model.go]: Implemented this logic, including that for `telemetry.distro.name`, in 4b9cca0bdf.  Would you mind taking a look at the unit test cases, particularly the ones in `TestEncodeLogECSModeAgentName` to make sure I got the logic right?  Thanks!

File: exporter/elasticsearchexporter/model.go
Comment (ycombinator) [reply exporter/elasticsearchexporter/model.go]: Implemented in d36a76fff8.

File: exporter/elasticsearchexporter/model.go
Comment (felixbarny) [reply exporter/elasticsearchexporter/model.go:273]: > One thing I think we should not carry across from apm-data is setting service.language.name when telemetry.sdk.language is unspecified. 

Sounds good. But maybe we have to adapt some of the assumptions we have in the APM UI about this field always being present.

> Maybe we can just skip setting service.language.name altogether for now, since it's included in agent.name?

The `service.language.name` attribute is used a lot in our UI. Since it directly translates to `service.language.name` I think I'd just add it to the `resourceAttrsConversionMap`.

Comment (axw) [reply exporter/elasticsearchexporter/model.go:273]: > The service.language.name attribute is used a lot in our UI. Since it directly translates to service.language.name I think I'd just add it to the resourceAttrsConversionMap.

This PR is focused on logs though, and I think the `service.language.name` usage in the (APM) UI is not relevant here. Could we start without it, and add it in later if there's a need?

Comment (felixbarny) [reply exporter/elasticsearchexporter/model.go:273]: let's do that

Comment (ycombinator) [reply exporter/elasticsearchexporter/model.go]: Removed code for setting `service.language.name` in cb5a3b88bb.

File: exporter/elasticsearchexporter/model.go
Comment (ycombinator) [reply exporter/elasticsearchexporter/model.go]: Implemented in 04a681bd49.

File: exporter/elasticsearchexporter/model.go
Comment (ycombinator) [reply exporter/elasticsearchexporter/model.go]: Implemented as part of 04a681bd49.

**Review by axw — APPROVED** [review]
LGTM!

**Comment by JaredTan95** [comment]
wait https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/31553 get merged.

**Review by felixbarny — APPROVED** [review]

File: .chloggen/exporter-elasticsarch-more-ecs-mode.yaml
Comment (carsonip) [inline .chloggen/exporter-elasticsarch-more-ecs-mode.yaml:13]:
```suggestion
issues: [31694]
```

**Review by carsonip — CHANGES_REQUESTED** [review]

**Review by carsonip — APPROVED** [review]

**Comment by ycombinator** [comment]
Hi @JaredTan95, I've rebased this PR on `main` and addressed conflicts after it's pre-requisite PR, #31553, was merged yesterday.  I've also incorporated the changes made in #31553 into the structure being used by the code in this PR (see 3533ef0cb094afbee112cb104178fb7659268da5).  

Would you mind re-reviewing this PR when you get a chance?  Thanks!

**Review by JaredTan95 — APPROVED** [review]
@ycombinator Thank you very much for your change and letting me know, I am glad to review, and changes LGTM.

**Comment by ycombinator** [comment]
@JaredTan95 The required CI check, https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/8641453852/job/23691083097?pr=31694, is failing on this PR but the failure seems unrelated to the changes in this PR:

> Failed while attempting to install arkade. Please manually install:
> ...

What is normally done in such cases?  Is there some way to re-run CI on this PR?

**Comment by ycombinator** [comment]
CI is [failing](https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/8780508502/job/24090919421?pr=31694#step:8:151) on an unrelated unit test that's also [failing](https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/8780289315/job/24090007607#step:8:152) on `main`.  @open-telemetry/collector-contrib-maintainer, could one of you please review this PR and merge if you think it's okay.  The codeowner, @JaredTan95, has already approved the PR.  Thanks!

**Review by carsonip — APPROVED** [review]

**Comment by JaredTan95** [comment]
> CI is [failing](https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/8780508502/job/24090919421?pr=31694#step:8:151) on an unrelated unit test that's also [failing](https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/8780289315/job/24090007607#step:8:152) on `main`. @open-telemetry/collector-contrib-maintainer, could one of you please review this PR and merge if you think it's okay. The codeowner, @JaredTan95, has already approved the PR. Thanks!
> 
> [EDIT] Here's the issue tracking the unrelated failing test: #32391

Thanks for continuing to follow up, but we should wait until the main branch fixes this to make sure all CI passes.

**Review by andrzej-stencel — APPROVED** [review]

**Comment by BenEfrati** [comment]
Hi @ycombinator , @andrzej-stencel 

There is a regression after this change - duplicate key in json:
before this change request to ELK was:
```
{
...
   "host":{
      "name":"hostname"
   },
   "os":{
      "description":"Windows 11 10.0",
      "type":"windows"
   },
...
}
```

after this change:
```
{
...
   "host":{
      "hostname":"hostname",
      "os":{
         "full":"Windows 11 10.0",
         "platform":"windows"
      }
   },
...
 "host":{
      "os":{
         "type":"windows"
      }
   },
...
}
```

in this example `host` key is duplicated, result in invalid json and error from ELK
```
 "error": {
  "reason": "failed to parse",
  "type": "mapper_parsing_exception"
  },
  "status": 400
```

Using same resource attributes:
```
Resource attributes:
host.name: Str(hostname)
os.description: Str(Windows 11 10.0)
os.type: Str(windows)
```
_this is from debug exporter_

File: exporter/elasticsearchexporter/model.go
Comment (BenEfrati) [inline exporter/elasticsearchexporter/model.go:336]: in case host is already part of document it'll result in duplicate host key in json request:
Resource attributes:
```
host.name: Str(hostname)
os.description: Str(Windows 11 10.0)
os.type: Str(windows)
```

**Pull request merged**
